File: pkg-r/tests/testthat/test-remote-clinvar.R
# Offline tests for clinvar remote mode. Existence is checked against NCBI
# E-utilities esearch; the biogate.remote_transport option replaces it.

.stub_clinvar <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}
.clinvar_hits <- function(n = 1) {
  list(esearchresult = list(count = as.character(n)))
}

test_that("the url builds the esearch endpoint", {
  url <- .remote_resolvers$clinvar$url(NULL, "VCV000012345")
  expect_identical(
    url,
    paste0(
      "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi",
      "?db=clinvar&term=VCV000012345&retmode=json"
    )
  )
})

test_that("an existing accession is valid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_clinvar(200, .clinvar_hits(1))
  )
  res <- check_id("VCV000012345", "clinvar", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "VCV000012345")
  expect_true(is.na(res$suggestion))
})

test_that("an absent accession is invalid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_clinvar(200, .clinvar_hits(0))
  )
  res <- check_id("VCV999999999", "clinvar", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_clinvar(200, .clinvar_hits(1))
  )
  res <- check_id("vcv000012345", "clinvar", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "VCV000012345")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("VCV12345", "clinvar", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_clinvar(500))
  expect_error(
    check_id("VCV000012345", "clinvar", how = "remote"),
    class = "biogate_error_remote"
  )
  expect_false(
    file.exists(.remote_cache_path("clinvar", "esearch", "VCV000012345"))
  )
})

File: pkg-py/tests/test_remote_clinvar.py
"""ClinVar remote mode: live existence checks via NCBI E-utilities esearch."""

import pytest

import biobouncer
import biobouncer._remote as remote
from biobouncer import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def _hits(n=1):
    return {"esearchresult": {"count": str(n)}}


def test_url_builds_the_esearch_endpoint():
    from biobouncer._registry import get_source

    url = remote._clinvar_url(get_source("clinvar"), "VCV000012345")
    assert url == (
        "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"
        "?db=clinvar&term=VCV000012345&retmode=json"
    )


def test_existing_accession_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200, _hits(1)))
    res = biobouncer.check_id("VCV000012345", source_db="clinvar", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "VCV000012345"
    assert res.suggestion is None


def test_absent_accession_is_invalid(monkeypatch):
    # esearch answers 200 with a zero count for an accession with no hits.
    monkeypatch.setattr(remote, "_http_get", _stub(200, _hits(0)))
    res = biobouncer.check_id("VCV999999999", source_db="clinvar", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200, _hits(1)))
    res = biobouncer.check_id("vcv000012345", source_db="clinvar", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "VCV000012345"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biobouncer.check_id("VCV12345", source_db="clinvar", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biobouncer.check_id("VCV000012345", source_db="clinvar", how="remote")
    path = remote._remote_cache_path("clinvar", "esearch", "VCV000012345")
    assert not path.is_file()
